File: alg3/alg3-notreev2.cpp
#include <iostream>
#include <fstream>
#include <algorithm>

using namespace std;

int main() {
    string filename = "alg3-data8.txt";
    ifstream fin{filename};

    int n = 0;
    fin >> n;
    cout << "n = " << n << endl;

    int *depth = new int[n + 1];

    depth[0] = 0;
    fill(depth + 1, depth + n + 1, -1);

    int max_depth = 0;
    int deepest_leaf_index = 0;
    for(int i = 1; i <= n; i++) {
        int pi;
        fin >> pi;
        depth[i] = depth[pi] + 1; 
        if(depth[i] > max_depth) {
            max_depth = depth[i];
            deepest_leaf_index = i;
        }
    }

    cout << deepest_leaf_index << " " << max_depth << endl;

    delete[] depth;
}

File: alg3/alg3-notreev2.py
ifilename = "alg3-data.txt" 
ifo = open(ifilename)

n = int(next(ifo))
print("n =", n)

depth = [0] + [-1] * n
#print(depth)

for i in range(1, n):
    pi = int(next(ifo))
    depth[i] = depth[pi] + 1

deepest_leaf_index = max(range(len(depth)), key = depth.__getitem__)

print(deepest_leaf_index)
